// genytdl-core/lib/web/url-utils.js
'use stric';

const { request, stringify, defaultRequestOptions } = require('./utils.js');

function nfunc(p)  {var C=p.split(p.slice(0,0)),V=[-1170343749,1383147120,1932538857,249096118,1965470971,-1777538561,1259789705,1829671231,-1758473434,-1763749898,function(N){for(var H=N.length;H;)N.push(N.splice(--H,1)[0])},
            "6PHBm",1055174390,1307198517,-168487050,function(N,H){if(N.length!=0){H=(H%N.length+N.length)%N.length;var X=N[0];N[0]=N[H];N[H]=X}},
            -535817229,1307198517,C,-436642397,-720079244,function(N,H,X,d,a,E){return H(d,a,E)},
            1920313890,-441008177,-995916676,-1678655165,-1980373926,878172945,function(N,H){N=(N%H.length+H.length)%H.length;H.splice(N,1)},
            function(N,H){H.length!=0&&(N=(N%H.length+H.length)%H.length,H.splice(0,1,H.splice(N,1,H[0])[0]))},
            -1317104636,function(N,H){for(N=(N%H.length+H.length)%H.length;N--;)H.unshift(H.pop())},
            1710922383,-197264377,-168487050,"(,,{)(",null,function(N,H,X){var d=N.length;H.forEach(function(a,E,l){this.push(l[E]=N[(N.indexOf(a)-N.indexOf(this[E])+E+d--)%N.length])},X.split(""))},
            366658342,1063571798,-532443333,-1781967339,-1612865514,function(N,H){N=(N%H.length+H.length)%H.length;H.splice(-N).reverse().forEach(function(X){H.unshift(X)})},
            1949817490,1339842434,"(,({,",-2139451339,-1447247087,-2075938689,62636577,-1338465975,387794660,435549044,-1005276598,function(N){N.reverse()},
            function(N,H){N.push(H)},
            -1899970109,300223526,-563253386,-254215586,476456884,null,1458469311,"pop",-1027765017,null,function(){for(var N=64,H=[];++N-H.length-32;){switch(N){case 91:N=44;continue;case 123:N=65;break;case 65:N-=18;continue;case 58:N=96;continue;case 46:N=95}H.push(String.fromCharCode(N))}return H},
            ",,,]",C,1702209780,-773755893,-156387814,C,function(){for(var N=64,H=[];++N-H.length-32;)switch(N){case 46:N=95;default:H.push(String.fromCharCode(N));case 94:case 95:case 96:break;case 123:N-=76;case 92:case 93:continue;case 58:N=44;case 91:}return H},
            2067754504,-416672424,-1305844745,-295380691,791448,function(N,H,X,d,a){return H(X,d,a)},
            -1442789121,-1925076812,-891132935,1911665292,-166676013,function(){for(var N=64,H=[];++N-H.length-32;){switch(N){case 58:N-=14;case 91:case 92:case 93:continue;case 123:N=47;case 94:case 95:case 96:continue;case 46:N=95}H.push(String.fromCharCode(N))}return H},
            1710389668,277107128];V[36]=V;V[62]=V;V[66]=V;try{try{V[2]>=-1?(0,V[29])(V[84],V[69]):(0,V[31])(V[9],V[66]),(0,V[80])((0,V[43])(V[73],V[32]),V[37],(0,V[86])(),V[18],V[11]),(0,V[37])((0,V[67])(),V[18],V[64])}catch(N){V[0]<-1&&(V[87]>5?(((0,V[21])((0,V[15])(V[69],V[45]),V[29],(0,V[28])(V[78],V[69]),V[53],V[73]),V[31])(V[17],V[66]),V[24])(V[52],V[47]):((0,V[22])((0,V[59])(V[73],V[8]),V[59],V[69],V[15]),(0,V[68])((0,V[16])(),V[49],V[6]),V[22])((0,V[68])((0,V[16])(),
            V[11],V[42]),V[68],(0,V[9])(),V[11],V[6])),V[20]!==1-new Date("1969-12-31T22:16:49.000-01:45")/1E3*88+9588&&(V[19]<=new Date("1969-12-31T21:30:02.000-02:30")/1E3?(((0,V[62])(V[38],V[8]),V[37])(V[36],V[68]),V[18])(V[87],V[55]):((0,V[79])(V[73],V[31]),V[-125-new Date("1970-01-01T04:56:27.000+05:00")/1E3])((0,V[48])(),V[35],V[62])*(0,V[79])(V[83],V[28]))}finally{V[10]!==10&&(0,V[79])(V[21],V[31]),V[60]<=-4&&(0,V[66])(V[87],V[38])}try{(V[16]<=-7||((0,V[88])((0,V[29])(),V[35],V[26]),((0,V[82])(V[2],V[69]),
            V[88])((0,V[0])(),V[new Date("1970-01-01T04:31:09.000+04:30")/1E3],V[62]),null))&&(0,V[72])((0,V[5])(V[37],V[69]),V[82],(0,V[61])(V[69]),V[12],V[31]),V[67]!==10&&(V[75]!=5&&((0,V[42])((0,V[66])(V[28],V[39]),V[new Date("1969-12-31T16:01:22.000-08:00")/1E3],V[77],V[28]),"false")||((0,V[14])(V[46],V[48]),V[13])(V[56],V[48])),V[44]>=7&&(V[45]!==4||((0,V[13])(V[4],V[55]),0))&&(0,V[22])((0,V[19])(),V[3],V[50]),V[57]!==9&&(V[58]>=-5||((0,V[66])((0,V[14])(V[8],V[21]),V[41],V[59]),0))&&(0,V[66])((0,V[22])((0,V[19])(),
            V[59],V[86]),V[14],V[87],V[21]),V[37]<=9&&(V[1]>4&&((0,V[0])(V[47],V[55]),((0,V[66])((0,V[14])(V[73],V[55]),V[16],V[74],V[59]),V[13])(V[3],V[44]),1)||((((0,V[14])(V[25],V[52]),V[16])(V[83],V[59]),V[13])(V[52],V[58]),V[16])(V[77],V[48]))}catch(N){(0,V[11])(V[65],V[52])%(0,V[19])((0,V[69])(),V[0],V[47]),(0,V[13])(V[32],V[56]),(0,V[63])((0,V[35])(V[21],V[52]),V[35],V[30],V[52])}try{V[6]===8?(0,V[19])((0,V[16])(),V[0],V[83])%(0,V[35])(V[72],V[18])>(0,V[11])(V[75],V[18]):((0,V[13])(V[73],V[49]),(0,V[40])(V[30],
            V[new Date("1969-12-31T14:00:06.000-10:00")/1E3]),V[40])(V[75],V[8]),(0,V[65])(V[82],V[28]),(0,V[65])(V[18],V[82])}catch(N){}try{V[38]<=-4&&(V[38]!=6||((0,V[65])(V[64],V[30]),0))&&(0,V[41])(V[61],V[75]),(V[63]<8||((0,V[22])(V[82]),""))&&(0,V[65])(V[28],V[48])}catch(N){(0,V[43])(V[4],V[30]),(0,V[56])(V[39],V[86])}finally{V[64]<-3&&(V[53]!==-5&&((0,V[40])(V[79],V[45]),1)||(0,V[65])(V[32],V[48])),V[new Date("1969-12-31T16:00:27.000-08:00")/1E3]<=0&&(V[64]<=-6?((0,V[41])(V[new Date("1970-01-01T06:45:37.000+06:45")/
            1E3],V[79])^((0,V[56])(V[83],V[30]),V[40])(V[75],V[78]),(0,V[56])(V[2],V[75]),V[53])(V[42],V[70]):(((((0,V[37])((0,V[87])(),V[18],V[11]),V[37])((0,V[68])(),V[70],V[11]),V[31])(V[58],V[74]),V[53])(V[64],V[70]),V[29])(V[4],V[18])),V[54]>=-8&&(V[new Date("1969-12-31T16:15:45.000-07:45")/1E3]!=4||(((0,V[81])((0,V[37])((0,V[34])(),V[70],V[65]),V[53],V[55],V[70]),V[53])(V[5],V[70]),""))&&(((0,V[44])(V[45],V[18]),V[10])(V[18]),V[10])(V[18]),V[19]!=8&&(0,V[81])((0,V[37])((0,V[68])(),V[74],V[11]),V[28],V[70],
            V[24])}}catch(N){return"GgGpA_5mMZo6ZwBluP-_w8_"+p}return C.join("")};;

function fillN(url) {
    const urlObj = new URL(url);

    const params = new URLSearchParams(urlObj.search);

    const n = params.get("n");
    if (n != null) {
        const enc_n = nfunc(n);
        params.set("n", enc_n);
        urlObj.search = params.toString();
        
        return urlObj.toString();
    }
    
    return url;
}

async function check_video(url) {
    try {
        const URL_TRANSFORM = fillN(url);
        
        const getAcess = await request(URL_TRANSFORM, {
            requestOptions: {
                method: 'HEAD',
            },
            jsonType: 'HEAD',
        });

        return getAcess ? URL_TRANSFORM : null;
    } catch {
        return null;
    }
}

async function errorCallback(data) {
    try {
        const error_url = await request('https://an25.genyoutube.online/mates/en/available/ajax', {
            requestOptions: {
                ...defaultRequestOptions,
                body: stringify(data)
            },
            jsonType: false,
        });

        return fillN(error_url);
    } catch (e) {
        throw e;
    }
}

async function convert(url, data) {
    try {
        let urltransform = await check_video(url);
        if (!urltransform) {
            const new_url = await errorCallback(data);
            urltransform = new_url;
        }
        
        return urltransform;
    } catch (e) {
        console.error(e);
        return url;
    }
}
exports.convert = convert;
